cpp/src/Slice/Parser.cpp

```diff
@@ -4575,19 +4592,25 @@ Slice::DataMember::DataMember(
 // ----------------------------------------------------------------------
 
 UnitPtr
-Slice::Unit::createUnit(bool all, const StringList& defaultFileMetadata)
+Slice::Unit::createUnit(string_view languagePrefix, bool all, const StringList& defaultFileMetadata)
 {
     MetadataList defaultMetadata;
     for (const auto& metadataString : defaultFileMetadata)
     {
         defaultMetadata.push_back(make_shared<Metadata>(metadataString, "<command-line>", 0));
     }
 
-    UnitPtr unit{new Unit{all, std::move(defaultMetadata)}};
+    UnitPtr unit{new Unit{languagePrefix, all, std::move(defaultMetadata)}};
```

Comment:
The code above is a mystery to me. Would be nice to add a comment.

Reply:
This is not my code, but as I read it, we're creating a cyclic `shared_ptr` here.
Which is just leaking memory, right?

Reply:
SyntaxTreeBase has a _unit, and Unit is a SyntaxTreeBase so it has a unit.

We can consider replace this by a virtual `unit()` method, which for Unit can return `shared_from_this`

Reply:
I think another option is just not having `Unit` inherit from `SyntaxTreeBase`.
It's a pretty small API surface, most of which isn't useful for `Unit`:
```
virtual void destroy();                   // Only sets `_unit = nullptr;`
UnitPtr unit() const;                     // Useless on `Unit`, it already has access to itself.
DefinitionContextPtr definitionContext() const;     // Always `nullptr` for `Unit`
virtual void visit(ParserVisitor* visitor);         // Only useful function
```
So, it'd actually be nice to remove this useless inheritance, or simplify it somehow.

Unfortunately, we use `SyntaxTreeBasePtr` as a base type for many places, so this would require a good deal of reshuffling.

Reply:
I was talking about this code:
```cpp
    MetadataList defaultMetadata;
    for (const auto& metadataString : defaultFileMetadata)
    {
        defaultMetadata.push_back(make_shared<Metadata>(metadataString, "<command-line>", 0));
    }
```